File: apps/android/core/testing/src/main/kotlin/app/gauja/core/testing/CachedTitleRenderer.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.testing

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.widthIn
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.unit.dp
import app.gauja.core.common.AppError
import app.gauja.core.designsystem.GaujaTheme
import app.gauja.core.model.media.MediaType
import app.gauja.core.model.media.TitleSummary
import app.gauja.core.model.media.TmdbId
import app.gauja.core.model.servers.Cached
import app.gauja.core.ui.media.TitleCard
import app.gauja.core.ui.state.ContentMessage
import java.time.Instant

@Composable
fun CachedTitleRenderer(
    title: TitleSummary,
    modifier: Modifier = Modifier,
    poster: Painter? = null,
) {
    CachedTitleRenderer(CachedTitleState.Offline(Cached(title, Instant.EPOCH)), modifier, poster)
}

@Composable
fun CachedTitleRenderer(
    state: CachedTitleState,
    modifier: Modifier = Modifier,
    poster: Painter? = null,
    dark: Boolean = true,
    onOpen: (MediaType, TmdbId) -> Unit = { _, _ -> },
) {
    GaujaTheme(darkTheme = dark) {
        Column(modifier.widthIn(max = 220.dp)) {
            when (state) {
                CachedTitleState.Loading -> CircularProgressIndicator()
                CachedTitleState.Empty -> Text("No cached titles")
                CachedTitleState.Denied -> ContentMessage(AppError.PERMISSION)
                is CachedTitleState.Failed -> ContentMessage(state.error, onRetry = {})
                is CachedTitleState.Offline -> {
                    Text("Offline · cached ${state.title.fetchedAt}")
                    TitleCard(state.title.value, onOpen, poster = poster)
                }
            }
        }
    }
}

File: apps/ios/Packages/Testing/Sources/GaujaTesting/CachedTitleRenderer.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
#if canImport(UIKit)
    import Common
    import DesignSystem
    import Model
    import SwiftUI
    import UI

    import UIKit

    public enum CachedTitleRenderer {
        public enum State {
            case loading, empty, denied
            case failed(AppError)
            case offline(Cached<TitleSummary>)
        }

        public static func contentHeight(_ title: TitleSummary, largeText: Bool) -> CGFloat {
            let content = GaujaTheme {
                TitleCard(title: title) { _, _ in }.frame(width: 220)
                    .environment(\.dynamicTypeSize, largeText ? .accessibility5 : .large)
            }
            let host = UIHostingController(rootView: content)
            host.traitOverrides.preferredContentSizeCategory = largeText ? .accessibilityExtraExtraExtraLarge : .large
            return host.sizeThatFits(in: CGSize(width: 220, height: CGFloat.greatestFiniteMagnitude)).height
        }

        public static func render(_ state: State, largeText: Bool = false, poster: CGImage? = nil) -> Bool {
            let content = GaujaTheme {
                CachedTitleContent(state: state, poster: poster).frame(width: 220)
                    .environment(\.dynamicTypeSize, largeText ? .accessibility5 : .large)
            }
            let host = UIHostingController(rootView: content)
            host.traitOverrides.preferredContentSizeCategory = largeText ? .accessibilityExtraExtraExtraLarge : .large
            let window = UIWindow(frame: CGRect(x: 0, y: 0, width: 240, height: 1600))
            window.rootViewController = host
            window.isHidden = false
            defer {
                window.isHidden = true
                window.rootViewController = nil
            }
            host.view.frame = window.bounds
            host.view.setNeedsLayout()
            host.view.layoutIfNeeded()
            var rendered = false
            let image = UIGraphicsImageRenderer(size: host.view.bounds.size).image { _ in
                rendered = host.view.drawHierarchy(in: host.view.bounds, afterScreenUpdates: true)
            }
            return rendered && image.cgImage?.width == Int(host.view.bounds.width * image.scale)
        }
    }

    private struct CachedTitleContent: View {
        let state: CachedTitleRenderer.State
        let poster: CGImage?
        var body: some View {
            VStack(alignment: .leading) {
                switch state {
                case .loading: ProgressView("Loading cached title")
                case .empty: Text("No cached titles")
                case .denied: ContentMessage(error: .permission)
                case .failed(let error): ContentMessage(error: error, retry: {})
                case .offline(let cached):
                    Text("Offline · cached \(cached.fetchedAt.formatted())")
                    TitleCard(title: cached.value, poster: poster.map { Image(decorative: $0, scale: 1) }) { _, _ in }
                }
            }
        }
    }

#endif
